server: introduce streaming ClusterFreeze endpoint

tbg opened the pull request:
Increase usability of the `freeze-cluster` command by responding with a JSON
stream which reflects the server-side polling of all Stores in the system.

This isn't extremely useful, but it was a good opportunity to clear the path
for future commands wishing to do the same.

```
$ ./cockroach freeze-cluster
proposed freeze to 5 ranges
waiting for 1 store apply operation
node 1, store 1: ready
ok
$ ./cockroach freeze-cluster --undo
proposed freeze to 5 ranges
waiting for 1 store apply operation
node 1, store 1: ready
ok
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7090)

<!-- Reviewable:end -->

tamird commented:
Looks pretty slick to me.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> #### server: introduce streaming ClusterFreeze endpoint
> 
> Increase usability of the `freeze-cluster` command by responding with a JSON
> 
> stream which reflects the server-side polling of all Stores in the system.
> 
> 
> 
> This isn't extremely useful, but it was a good opportunity to clear the path
> 
> for future commands wishing to do the same.
> 
> 
> 
> ```
> 
> $ ./cockroach freeze-cluster
> 
> proposed freeze to 5 ranges
> 
> waiting for 1 store apply operation
> 
> node 1, store 1: ready
> 
> ok
> 
> $ ./cockroach freeze-cluster --undo
> 
> proposed freeze to 5 ranges
> 
> waiting for 1 store apply operation
> 
> node 1, store 1: ready
> 
> ok
> 
> ```
> 
> 
> 
</details>


---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[acceptance/update_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgjXPK3cAWJ_nMuEin:-KJgjXPK3cAWJ_nMuEio:947219010) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/acceptance/update_test.go#L45)):*
> ```Go
> 
> 	var resp serverpb.ClusterFreezeResponse
> 	var m map[string]*serverpb.ClusterFreezeResponse
> ```

can you add a comment here explaining why this isn't a proto? we used to have a lot of json unmarshalling into maps rather than protos which I painstakingly removed; I'd hate for the precedent to be lost.

Actually, are you sure this can't be a struct?

Also, why does the value type need to be a pointer?

---

*[acceptance/update_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgjvLvIsCMG3DOulFI:-KJgjvLvIsCMG3DOulFJ:1451895011) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/acceptance/update_test.go#L112)):*
> ```Go
> 		"connection is closing",
> 		"deadline",
> 		"unknown field", // gRPC stream timeout JSON response
> ```

more words

---

*[cli/cli_test.go, line 866 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgk8Sq4HMOBA17YsSE:-KJgk8Sq4HMOBA17YsSF:-1598511654) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/cli_test.go#L866)):*
> ```Go
> 	assertOutput := func(msg string) {
> 		if !strings.HasSuffix(strings.TrimSpace(msg), "ok") {
> 			t.Fatal(msg)
> ```

expected ... to include ... 

---

*[cli/cli_test.go, line 875 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgk0oNvTn_14PwJPJc:-KJgk0oNvTn_14PwJPJd:-1121102124) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/cli_test.go#L875)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		assertOutput(out)
> ```

I think you should duplicate this code. If this ever fails it'll be impossible to tell which one failed.

---

*[cli/start.go, line 442 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgkEsJheoVQUoMYsXs:-KJgkEsJheoVQUoMYsXt:1870160137) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/start.go#L442)):*
> ```Go
> 		if err != nil {
> 			if err == io.EOF {
> 				break
> ```

nit: fmt.Println("ok"); return nil right here? doesn't matter i suppose.

---

*[server/admin.go, line 818 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgkigismcGxlxnkyik:-KJgkigismcGxlxnkyil:-1178468622) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/admin.go#L818)):*
> ```Go
> 							// Otherwise, forget that we tried the Store so that
> 							// the retry loop picks it up again.
> 							message += fmt.Sprintf("%d frozen, %d thawed ranges", resp.NumFrozen, resp.NumThawed)
> ```

nit: s/frozen/frozen ranges/

---

*[server/admin.go, line 875 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgks5ncm2tzrwiUiNf:-KJgks5ncm2tzrwiUiNg:-530397867) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/admin.go#L875)):*
> ```Go
> 		}
> 		if err := stream.Send(&serverpb.ClusterFreezeResponse{
> 			Message: fmt.Sprintf("waiting for %d store%s apply operation",
> ```

"to"?

---

*[server/admin_test.go, line 695 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgk_eKdBVOryHgrW8N:-KJgk_eLUDkF_uNncnJT:-1762184772) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/admin_test.go#L695)):*
> ```Go
> 
> 		var resp serverpb.ClusterFreezeResponse
> 		m := map[string]*serverpb.ClusterFreezeResponse{}
> ```

ditto regarding this map.

---

*[server/server.go, line 598 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgkUtYMoQf77Wy1V7J:-KJgkUtYMoQf77Wy1V7K:1208293375) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/server.go#L598)):*
> ```Go
> // which access streaming endpoints (as exercised by the acceptance tests
> // at time of writing).
> func (*gzipResponseWriter) Flush() {}
> ```

this thing has outlived its usefulness. I'm going to submit a PR to remove it.

---

*[util/http.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJglaKVzdcf6wVFvGkp:-KJglaKVzdcf6wVFvGkq:196407943) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/http.go#L117)):*
> ```Go
> 	decoder := marshaler.NewDecoder(resp.Body)
> 	for {
> 		err := decoder.Decode(dest)
> ```

nit this can be in the `if` below

---

*[util/jsonpb_marshal.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgl1lkmlAaTcgZI7kU:-KJgl1lkmlAaTcgZI7kV:-1066271383) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/jsonpb_marshal.go#L48)):*
> ```Go
> }
> 
> // a lower-case version of marshal allow for a call from marshalNonProtoField
> ```

to allow

---

*[util/jsonpb_marshal.go, line 68 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJglec2MTR-xrx0gh6R:-KJglec2MTR-xrx0gh6S:-1615833897) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/jsonpb_marshal.go#L68)):*
> ```Go
> }
> 
> // marshalNonProto marshals a non-message field of a protobuf message.
> ```

i think i'd remove this comment, but up to you

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-:-KJgjTF2IO6hnT6b7kuC:-628223118)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 5 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[acceptance/update_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgjXPK3cAWJ_nMuEin:-KJgtrBfpLEfd-LoTtdJ:1378081721) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/acceptance/update_test.go#L45)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you add a comment here explaining why this isn't a proto? we used to have a lot of json unmarshalling into maps rather than protos which I painstakingly removed; I'd hate for the precedent to be lost.
> 
> Actually, are you sure this can't be a struct?
> 
> Also, why does the value type need to be a pointer?
</details>
It's what grpc-gateway returns for streaming rpcs, it's something like

```js
{
  "result": {
    "ranges_affected": 5,
  }
}
```

and so I didn't have a chance to make this work better. The pointer is needed because the unmarshaling code ... checks for a pointer (or maybe it's required to implement `proto.Message`, something like that).

---

*[acceptance/update_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgjvLvIsCMG3DOulFI:-KJgvKO-_aUC81Zkdx8-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/acceptance/update_test.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> more words
</details>
Done.

---

*[cli/cli_test.go, line 866 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgk8Sq4HMOBA17YsSE:-KJgsdKuTcsHYR-dR_sP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/cli_test.go#L866)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> expected ... to include ... 
</details>
Done.

---

*[cli/cli_test.go, line 875 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgk0oNvTn_14PwJPJc:-KJgsdiNQikbj230JXwA:-1248716367) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/cli_test.go#L875)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think you should duplicate this code. If this ever fails it'll be impossible to tell which one failed.
</details>
Used `ErrorfSkipFrames` in the assertion.

---

*[cli/start.go, line 442 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgkEsJheoVQUoMYsXs:-KJgsnFjDWuYbKugslLH:-1198645626) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/cli/start.go#L442)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: fmt.Println("ok"); return nil right here? doesn't matter i suppose.
</details>
Yeah, I'll leave as is. `return` at the end is soothing to the eye.

---

*[server/admin.go, line 818 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgkigismcGxlxnkyik:-KJgtBsObYv2aFIS3qxo:1869449973) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/admin.go#L818)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: s/frozen/frozen ranges/
</details>
"x frozen and %d thawed ranges" ok?

---

*[server/admin.go, line 875 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgks5ncm2tzrwiUiNf:-KJgtKhst3PrfjHwiWg5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/server/admin.go#L875)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "to"?
</details>
Done.

---

*[util/http.go, line 117 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJglaKVzdcf6wVFvGkp:-KJgtULXmAuiz6UQOMeB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/http.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit this can be in the `if` below
</details>
Done.

---

*[util/jsonpb_marshal.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgl1lkmlAaTcgZI7kU:-KJgthyBdA_D34TJd41_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/jsonpb_marshal.go#L48)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> to allow
</details>
Done.

---

*[util/jsonpb_marshal.go, line 68 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJglec2MTR-xrx0gh6R:-KJgtkTYu1YGXkw99Ddi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/util/jsonpb_marshal.go#L68)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think i'd remove this comment, but up to you
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[acceptance/update_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgjXPK3cAWJ_nMuEin:-KJgw1fwW9S0imL387-Q:845671129) ([raw file](https://github.com/cockroachdb/cockroach/blob/8675cf66dbe6ddaa5e1e0a4698f67ed7bac589a6/acceptance/update_test.go#L45)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It's what grpc-gateway returns for streaming rpcs, it's something like
> 
> ```js
> {
>   "result": {
>     "ranges_affected": 5,
>   }
> }
> ```
> 
> and so I didn't have a chance to make this work better. The pointer is needed because the unmarshaling code ... checks for a pointer (or maybe it's required to implement `proto.Message`, something like that).
</details>
Just asking for the map to be a struct instead, as discussed in person.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Added some reflection, PTAL.

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Looks pretty slick to me.
</details>


---

Review status: 8 of 14 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-:-KJiX3bnwTRMFPlee8tm:-1274245127)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[util/http.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjjq_3LNFnSkfDhX2p:-KJjjq_3LNFnSkfDhX2q:-784756338) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L97)):*
> ```Go
> 	path string,
> 	request proto.Message,
> 	dest interface{},
> ```

i think it should be possible to make this `proto.Message`. you'll just need to remove the `PtrTo` bit.

---

*[util/http.go, line 118 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjj_eLrK_ky3UNpOdm:-KJjj_eLrK_ky3UNpOdn:-1906665971) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L118)):*
> ```Go
> 	// we must pass a *map[string]*concreteType or it won't work (in
> 	// particular, using a struct or replacing `*concreteType` with either
> 	// `concreteType` or `proto.Message leads to errors). This method should do
> ```

s/Message/Message`/

---

*[util/http.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjkUDnsX0vTXyEbBIa:-KJjkUDnsX0vTXyEbBIb:1199207090) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L139)):*
> ```Go
> 			return Errorf("unexpected JSON response: %+v", m)
> 		}
> 		callback(v.Elem().Interface())
> ```

`callback(v.Interface().(proto.Message))` and then you can make the type nicer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on server/admin.go:
```diff
@@ -923,10 +940,16 @@ func (s *adminServer) ClusterFreeze(
-	return &resp, s.waitForStoreFrozen(stores, req.Freeze)
+	if err := stream.Send(&serverpb.ClusterFreezeResponse{
+		RangesAffected: totalAffected,
+		Message:        fmt.Sprintf("proposed freeze to %d ranges", totalAffected),
```
Why not change this message to say "unfreeze" or "thaw" instead of "freeze" in the event that `!req.Freeze`?

tschottdorf commented:
Review status: 9 of 14 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[server/admin.go, line 948 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJgcqMgaxLWkGGemHWK-r3-948:-KJkRa_SA3cfuYtu5CgZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/server/admin.go#L948)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Why not change this message to say "unfreeze" or "thaw" instead of "freeze" in the event that `!req.Freeze`?
</details>
Done.

---

*[util/http.go, line 97 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjjq_3LNFnSkfDhX2p:-KJkTRG6G9iHfj11ovy0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think it should be possible to make this `proto.Message`. you'll just need to remove the `PtrTo` bit.
</details>
Done.

---

*[util/http.go, line 118 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjj_eLrK_ky3UNpOdm:-KJkRj3aU6lr6j1gY2MI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L118)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Message/Message`/
</details>
Done.

---

*[util/http.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090#-KJjkUDnsX0vTXyEbBIa:-KJkTQkTLKsLL7-UYc6k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5d4bdf5b10a430b39e9d2bfafb7f9fda332aad/util/http.go#L139)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `callback(v.Interface().(proto.Message))` and then you can make the type nicer
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: introduce streaming ClusterFreeze endpoint

Increase usability of the `freeze-cluster` command by responding with a JSON
stream which reflects the server-side polling of all Stores in the system.

This isn't extremely useful, but it was a good opportunity to clear the path
for future commands wishing to do the same.

```
$ ./cockroach freeze-cluster
proposed freeze to 5 ranges
waiting for 1 store apply operation
node 1, store 1: ready
ok
$ ./cockroach freeze-cluster --undo
proposed freeze to 5 ranges
waiting for 1 store apply operation
node 1, store 1: ready
ok
```

tamird commented:
Reviewed 5 of 5 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7090)*
<!-- Sent from Reviewable.io -->
